inference.py

```diff
@@ -10,169 +10,127 @@
-from utils.CreateDataset import MetaSegmentationDataset
 from utils.verifications import add_background_to_num_class
-from utils.visualization import vis, vis_from_batch
 from mlflow import log_params, set_tracking_uri, set_experiment, start_run, log_artifact, log_metrics
 
 try:
     import boto3
 except ModuleNotFoundError:
     pass
 
-colors = {0: [0, 0, 0],  # Background
-          1: [27, 120, 55],  # Vegetation
-          2: [116, 173, 209], # Hydro
-          3: [223, 194, 125],  # Roads
-          4: [150, 150, 150]}  # Buildings     
 
 @torch.no_grad()
-def segmentation_with_smoothing(raster, clip_gpkg, model, sample_size, overlap, num_bands, device):
+def segmentation(img_array, input_image, label_arr, num_classes, gpkg_name, model, sample_size, num_bands, device):
+
     # switch to evaluate mode
     model.eval()
-    img_array, input_image, dataset_nodata = image_reader_as_array(input_image=raster,
-                                                                   clip_gpkg=clip_gpkg)
-    metadata = add_metadata_from_raster_to_sample(img_array,
-                                                    input_image,
-                                                    meta_map=None,
-                                                    raster_info=None)
-    h, w, bands = img_array.shape
-    assert num_bands <= bands, f"Num of specified bands is not compatible with image shape {img_array.shape}"
-    if num_bands < bands:
-       img_array = img_array[:, :, :num_bands]
+    transforms = tta.Compose([tta.HorizontalFlip(), ])
 
-    padding = int(round(sample_size * (1 - 1.0/overlap)))
-    padded_img = pad(img_array, padding=padding, fill=0)
-    WINDOW_SPLINE_2D = _window_2D(window_size=sample_size, power=1)
-    WINDOW_SPLINE_2D = np.moveaxis(WINDOW_SPLINE_2D, 2, 0)
-    step = int(sample_size/overlap)
-    h_, w_ = padded_img.shape[:2]
-    pred_img = np.empty((h_, w_), dtype=np.uint8)
-    for row in tqdm(range(0, h_ - sample_size + 1, step), position=1, leave=False, desc='Inferring rows'):
-        with tqdm(range(0, w_ - sample_size + 1, step), position=2, leave=False, desc='Inferring columns') as _tqdm:
-            for col in _tqdm:
-                sample = {'sat_img': None, 'metadata': None}
-                sample['metadata'] = metadata
-                totensor_transform = augmentation.compose_transforms(
-                    params, dataset="tst", type='totensor'
-                )
-                sub_images = padded_img[row:row+sample_size, col:col+sample_size, :]
-                sample['sat_img'] = sub_images
-                sample = totensor_transform(sample)
-                inputs = sample['sat_img'].unsqueeze_(0)
-                inputs = inputs.to(device)
+    WINDOW_SPLINE_2D = _window_2D(window_size=sample_size, power=2.0)
+    WINDOW_SPLINE_2D = torch.as_tensor(np.moveaxis(WINDOW_SPLINE_2D, 2, 0), ).type(torch.float)
+    WINDOW_SPLINE_2D = WINDOW_SPLINE_2D.to(device)
```

Comment:
is smoothing being done systematically now? is there any interest in being able to infer without it?

Reply:
I think it works well now, there are no artifacts introduced and it takes advantage of GPU resources.